sdk/client/src/verifying_client.rs

```diff
@@ -0,0 +1,516 @@
+        let trusted_state_store = TrustedStateStore::new(storage)?;
+        Ok(Self {
+            inner,
+            trusted_state_store: Arc::new(RwLock::new(trusted_state_store)),
+        })
+    }
+
+    pub fn new_with_state(
+        inner: Client,
+        trusted_state: TrustedState,
+        storage: Box<dyn Storage>,
+    ) -> Self {
+        let trusted_state_store = TrustedStateStore::new_with_state(trusted_state, storage);
+        Self {
+            inner,
+            trusted_state_store: Arc::new(RwLock::new(trusted_state_store)),
+        }
+    }
+
+    /// Get a snapshot of our current trusted ledger [`Version`].
+    pub fn version(&self) -> Version {
+        self.trusted_state_store.read().unwrap().version()
+    }
+
+    /// Get a snapshot of our current trusted [`Waypoint`].
+    pub fn waypoint(&self) -> Waypoint {
+        self.trusted_state_store.read().unwrap().waypoint()
+    }
+
+    /// Get a snapshot of our current [`TrustedState`].
+    pub fn trusted_state(&self) -> TrustedState {
+        self.trusted_state_store
+            .read()
+            .unwrap()
+            .trusted_state()
+            .clone()
+    }
+
+    /// Issue `get_state_proof` requests until we successfully sync to the remote
+    /// node's current version (unless we experience a verification error or other
+    /// I/O error).
+    pub async fn sync(&self) -> Result<()> {
+        // TODO(philiphayes): retries
+        while self.sync_one_step().await? {}
+        Ok(())
+    }
+
+    /// Issue a single `get_state_proof` request and try to verify it. Returns
+    /// `Ok(true)` if, after verification, we still need to sync more. Returns
+    /// `Ok(false)` if we have finished syncing.
+    pub async fn sync_one_step(&self) -> Result<bool> {
+        let (state_proof_view, state) = self
+            .inner
+            .get_state_proof(self.version())
+            .await?
+            .into_parts();
+        let state_proof = StateProof::try_from(&state_proof_view)?;
+
+        let latest_li = state_proof.0.ledger_info();
+        ensure!(latest_li.version() == state.version);
+        ensure!(latest_li.timestamp_usecs() == state.timestamp_usecs);
+
+        self.verify_and_ratchet(&state_proof)?;
+
+        Ok(state_proof.1.more)
+    }
+
+    /// Verify and ratchet forward our trusted state using a state proof.
+    pub fn verify_and_ratchet(&self, state_proof: &StateProof) -> Result<()> {
+        let (latest_li, epoch_change_proof, _) = state_proof;
+
+        let trusted_state = self.trusted_state();
+        let change = trusted_state.verify_and_ratchet(latest_li, epoch_change_proof)?;
+
+        if let Some(new_state) = change.new_state() {
+            self.trusted_state_store
+                .write()
+                .unwrap()
+                .ratchet(new_state)?;
+        }
+
+        // TODO(philiphayes): verify the accumulator consistency proof; no-one
+        // seems to actually verify this anywhere?
```

Comment:
Could it be because if we verify the `EpochChangeProof`s, then we trust the current set of validators and then just need to verify the current `AccumulatorProof`s?

That being said, we are verifying `AccumulatorExtensionProof` in the consensus' safety rules currently, and I think it might be to achieve the same kind of result.   
But maybe it isn't and we should also be checking these accumulator consistency proofs in our safety rules??
@zekun000 maybe you can give us some insights?

Reply:
ExtensionProof includes the previous tree hashes so it's self-contained, iiuc ConsistencyProof only gives newly added leaves so the client needs to maintain the previous tree in order to verify it

Reply:
cc @msmouse I have some questions about the accumulator consistency proof.

1. How does verifying the consistency proofs fit in with the light client sync? For example, if we start at an epoch change `Waypoint`, we only have the accumulator root hash and version but no subtree roots, so we can't verify any consistency proofs without the remote also providing the subtree roots (I believe (?)).

2. Another issue I foresee is that the consistency proof returned from `get_state_proof(start_version)` takes us from `start_version --> remote's latest_li.version`, but we may only receive a partial epoch change proof (if we're more than 100 epochs behind). Our current `TrustedState::verify_and_ratchet` logic will (correctly) move us up to the end of the epoch change proof but not to the latest_li; however, we can't use the consistency proof to verify `start_version --> last(epoch_change_proof).version` (I believe (?)). Do we need a consistency proof for the epoch change proof specifically (or even for each LIwS in the epoch change proof)?

3. Should light clients do an initial sync from `start_version=0 --> waypoint.version` to build up their accumulator so they can verify these consistency proofs? Is this something we should be doing anyway?

3. Could we clarify the threat model a little? What does verifying the consistency proof guarantee us? Is it so light clients can get another alternate signal (apart from majority validator signatures) that the chain is functioning correctly? If I had to speculate, I would imagine one failure case it prevents could be a code bug leading to 2f+1 validators agreeing to a garbage/inconsistent root hash; in this case, clients would correctly halt at last verifiable state. One case where it doesn't seem to help might be 2f+1 malicious validators agreeing to an alternate tx history; in this case, the malicious validators could instead agree to a big `WriteSet` and accomplish the same result w/o halting clients.

Thanks : )

Reply:
1. Yes. I think the idea is the light client asks the frozen subtrees (i.e. a InMemoryAccumulator) set from the server initially and keep it maintained along the way by updating it with the consistency proof. 

2. I think we need to loop and get the full set of epoch change proofs, the proofs and a single consistency proof will move us to the next LI. Sounds good?

3. Yes, something like that. Basically this is to ask for the initial frozen subtrees, right? To be clear, you don't need to sync the entire accumulator, but just get the frozen subtree roots (The InMemoryAccumulator)

4. I think it tells you even if there was a hard fork, at least you and the latest LI you are receiving are on the same fork.. not much info, I'd admit. @wqfish  Am I correct?